Flaky test: TestSaramaConsumerWrapper_start_Messages

https://travis-ci.org/jaegertracing/jaeger/jobs/487803410
```
=== RUN   TestSaramaConsumerWrapper_start_Messages
2019-02-02T09:46:32.154Z	DEBUG	consumer/deadlock_detector.go:151	Starting global deadlock detector
2019-02-02T09:46:32.155Z	INFO	consumer/consumer.go:77	Starting main loop
2019-02-02T09:46:32.155Z	INFO	consumer/consumer.go:162	Starting error handler	{"partition": 316}
2019-02-02T09:46:32.155Z	INFO	consumer/consumer.go:111	Starting message handler	{"partition": 316}
2019-02-02T09:46:32.156Z	DEBUG	consumer/consumer.go:134	Got msg	{"msg": {"Key":null,"Value":null,"Topic":"morekuzambu","Partition":316,"Offset":1,"Timestamp":"0001-01-01T00:00:00Z","BlockTimestamp":"0001-01-01T00:00:00Z","Headers":null}}
2019-02-02T09:46:32.156Z	INFO	consumer/processor_factory.go:62	Creating new processors	{"partition": 316}
2019-02-02T09:46:32.157Z	DEBUG	processor/parallel_processor.go:50	Spawning goroutines to process messages	{"num_routines": 1}
2019-02-02T09:46:32.158Z	INFO	consumer/consumer.go:155	Closing partition consumer	{"partition": 316}
2019-02-02T09:46:32.158Z	INFO	consumer/consumer.go:131	Message channel closed. 	{"partition": 316}
2019-02-02T09:46:32.158Z	DEBUG	processor/parallel_processor.go:75	Initiated shutdown of processor goroutines
2019-02-02T09:46:32.158Z	INFO	processor/parallel_processor.go:78	Completed shutdown of processor goroutines
2019-02-02T09:46:32.158Z	INFO	consumer/consumer.go:173	Finished handling errors	{"partition": 316}
2019-02-02T09:46:32.158Z	INFO	consumer/consumer.go:158	Closed partition consumer	{"partition": 316}
2019-02-02T09:46:32.158Z	DEBUG	consumer/deadlock_detector.go:195	Closing deadlock detector	{"partition": 316}
2019-02-02T09:46:32.158Z	INFO	consumer/consumer.go:155	Closing partition consumer	{"partition": 316}
2019-02-02T09:46:32.158Z	INFO	consumer/deadlock_detector.go:115	Closing ticker routine	{"partition": 316}
2019-02-02T09:46:32.159Z	INFO	consumer/consumer.go:158	Closed partition consumer	{"partition": 316}
2019-02-02T09:46:32.159Z	DEBUG	consumer/deadlock_detector.go:179	Closing all partitions deadlock detector
2019-02-02T09:46:32.159Z	INFO	consumer/consumer.go:106	Closing parent consumer
--- FAIL: TestSaramaConsumerWrapper_start_Messages (0.01s)
    <autogenerated>:1: PASS:	Process(*consumer.saramaMessageWrapper)
	Error Trace:	metricstest.go:49
			metricstest.go:37
			consumer_test.go:167
	Error:      	Not equal: 
	            	expected: int(0)
	            	actual  : int64(1)
	Test:       	TestSaramaConsumerWrapper_start_Messages
	Messages:   	expected metric name: sarama-consumer.partitions-held, tags: map[]
2019-02-02T09:46:32.161Z	DEBUG	consumer/deadlock_detector.go:159	Closing global ticker routine
```

Getting different stack trace now:
```
=== RUN   TestSaramaConsumerWrapper_start_Messages
2020-05-11T18:49:25.813Z	DEBUG	consumer/deadlock_detector.go:151	Starting global deadlock detector
2020-05-11T18:49:25.814Z	INFO	consumer/consumer.go:78	Starting main loop
2020-05-11T18:49:25.814Z	INFO	consumer/consumer.go:112	Starting message handler	{"partition": 316}
2020-05-11T18:49:25.815Z	DEBUG	consumer/consumer.go:141	Got msg	{"msg": {"Headers":null,"Timestamp":"0001-01-01T00:00:00Z","BlockTimestamp":"0001-01-01T00:00:00Z","Key":null,"Value":null,"Topic":"morekuzambu","Partition":316,"Offset":1}}
2020-05-11T18:49:25.815Z	INFO	consumer/processor_factory.go:65	Creating new processors	{"partition": 316}
2020-05-11T18:49:25.816Z	DEBUG	processor/parallel_processor.go:50	Spawning goroutines to process messages	{"num_routines": 1}
2020-05-11T18:49:25.817Z	INFO	consumer/consumer.go:169	Starting error handler	{"partition": 316}
    TestSaramaConsumerWrapper_start_Messages: consumer_test.go:161: PASS:	Process(consumer.saramaMessageWrapper)
2020-05-11T18:49:25.818Z	INFO	consumer/consumer.go:162	Closing partition consumer	{"partition": 316}
2020-05-11T18:49:25.818Z	INFO	consumer/consumer.go:138	Message channel closed. 	{"partition": 316}
2020-05-11T18:49:25.818Z	DEBUG	processor/parallel_processor.go:75	Initiated shutdown of processor goroutines
2020-05-11T18:49:25.818Z	INFO	processor/parallel_processor.go:78	Completed shutdown of processor goroutines
2020-05-11T18:49:25.818Z	DEBUG	consumer/deadlock_detector.go:195	Closing deadlock detector	{"partition": 316}
2020-05-11T18:49:25.818Z	INFO	consumer/consumer.go:162	Closing partition consumer	{"partition": 316}
2020-05-11T18:49:25.819Z	INFO	consumer/deadlock_detector.go:115	Closing ticker routine	{"partition": 316}
2020-05-11T18:49:25.819Z	INFO	consumer/consumer.go:165	Closed partition consumer	{"partition": 316}
2020-05-11T18:49:25.819Z	INFO	consumer/consumer.go:165	Closed partition consumer	{"partition": 316}
2020-05-11T18:49:28.815Z	PANIC	consumer/deadlock_detector.go:71	No messages processed in the last check interval	{"partition": -1, "stack": "goroutine 25 [running]:\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.newDeadlockDetector.func1(0xc0ffffffff)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:73 +0x2aa\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*deadlockDetector).start.func1(0xc0000c6180, 0xc0000b0420)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:163 +0x2bf\ncreated by github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*deadlockDetector).start\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:152 +0x346\n\ngoroutine 1 [chan receive]:\ntesting.(*T).Run(0xc00012fe60, 0x104f8f5, 0x28, 0x1067820, 0x1)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1043 +0x699\ntesting.runTests.func1(0xc00012fe60)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1284 +0xa7\ntesting.tRunner(0xc00012fe60, 0xc00017bce8)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:991 +0x1ec\ntesting.runTests(0xc00000f040, 0x18c4c60, 0x15, 0x15, 0x0)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1282 +0x528\ntesting.(*M).Run(0xc000168100, 0x0)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1199 +0x300\nmain.main()\n\t_testmain.go:142 +0x338\n\ngoroutine 24 [semacquire]:\nsync.runtime_Semacquire(0xc0000b0588)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/runtime/sema.go:56 +0x42\nsync.(*WaitGroup).Wait(0xc0000b0580)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/sync/waitgroup.go:130 +0xd4\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).Close(0xc0000c60f0, 0xc00009c480, 0xf2dae0)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:103 +0x175\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.TestSaramaConsumerWrapper_start_Messages(0xc00009c480)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer_test.go:165 +0xad4\ntesting.tRunner(0xc00009c480, 0x1067820)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:991 +0x1ec\ncreated by testing.(*T).Run\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1042 +0x661\n\ngoroutine 26 [chan receive]:\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).Start.func1(0xc0000c60f0)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:79 +0x289\ncreated by github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).Start\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:77 +0x7c\n\ngoroutine 27 [semacquire]:\nsync.runtime_SemacquireMutex(0xc0000c61bc, 0x1100000000, 0x1)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/runtime/sema.go:71 +0x47\nsync.(*Mutex).lockSlow(0xc0000c61b8)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/sync/mutex.go:138 +0x1c1\nsync.(*Mutex).Lock(0xc0000c61b8)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/sync/mutex.go:81 +0x7d\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).handleMessages.func1(0xc0000c60f0, 0x117cb00, 0xc0000a6690, 0xc0000b0580)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:121 +0x9d\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).handleMessages(0xc0000c60f0, 0x117cb00, 0xc0000a6690)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:139 +0xbc5\ncreated by github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).Start.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:92 +0x204\n\ngoroutine 28 [semacquire]:\nsync.runtime_SemacquireMutex(0xc0000c61bc, 0x900000000, 0x1)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/runtime/sema.go:71 +0x47\nsync.(*Mutex).lockSlow(0xc0000c61b8)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/sync/mutex.go:138 +0x1c1\nsync.(*Mutex).Lock(0xc0000c61b8)\n\t/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/sync/mutex.go:81 +0x7d\ngithub.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).handleErrors(0xc0000c60f0, 0x13c, 0xc00008a7e0)\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:170 +0x1d3\ncreated by github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*Consumer).Start.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/consumer.go:93 +0x26b\n"}
github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.newDeadlockDetector.func1
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:71
github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*deadlockDetector).start.func1
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:163
panic: No messages processed in the last check interval
goroutine 25 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00011bd90, 0xc000168200, 0x2, 0x2)
	/home/travis/gopath/pkg/mod/go.uber.org/zap@v1.13.0/zapcore/entry.go:230 +0x711
go.uber.org/zap.(*Logger).Panic(0xc0000862a0, 0x105563f, 0x30, 0xc000168200, 0x2, 0x2)
	/home/travis/gopath/pkg/mod/go.uber.org/zap@v1.13.0/logger.go:225 +0x96
github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.newDeadlockDetector.func1(0xc0ffffffff)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:71 +0x4a5
github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*deadlockDetector).start.func1(0xc0000c6180, 0xc0000b0420)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:163 +0x2bf
created by github.com/jaegertracing/jaeger/cmd/ingester/app/consumer.(*deadlockDetector).start
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/consumer/deadlock_detector.go:152 +0x346
FAIL	github.com/jaegertracing/jaeger/cmd/ingester/app/consumer	3.026s
FAIL
make: *** [cover] Error 1
```

--- a/cmd/ingester/app/consumer/consumer.go
+++ b/cmd/ingester/app/consumer/consumer.go
@@ -50,10 +50,13 @@ type Consumer struct {
 	partitionMapLock    sync.Mutex
 	partitionsHeld      int64
 	partitionsHeldGauge metrics.Gauge
+
+	messagesDoneChan chan string
+	errorsDoneChan   chan string
+	doneWg           sync.WaitGroup
 }
 
 type consumerState struct {
-	wg                sync.WaitGroup
 	partitionConsumer sc.PartitionConsumer
 }
 
@@ -68,6 +71,8 @@ func New(params Params) (*Consumer, error) {
 		deadlockDetector:    deadlockDetector,
 		partitionIDToState:  make(map[int32]*consumerState),
 		partitionsHeldGauge: partitionsHeldGauge(params.MetricsFactory),
+		messagesDoneChan:    make(chan string),
+		errorsDoneChan:      make(chan string),
 	}, nil
 }
 
@@ -78,50 +83,65 @@ func (c *Consumer) Start() {
 		c.logger.Info("Starting main loop")
 		for pc := range c.internalConsumer.Partitions() {
 			c.partitionMapLock.Lock()
-			if p, ok := c.partitionIDToState[pc.Partition()]; ok {
-				// This is a guard against simultaneously draining messages
-				// from the last time the partition was assigned and
-				// processing new messages for the same partition, which may lead
-				// to the cleanup process not completing
-				p.wg.Wait()
-			}
 			c.partitionIDToState[pc.Partition()] = &consumerState{partitionConsumer: pc}
-			c.partitionIDToState[pc.Partition()].wg.Add(2)
 			c.partitionMapLock.Unlock()
 			c.partitionMetrics(pc.Partition()).startCounter.Inc(1)
+
+			c.doneWg.Add(2)
 			go c.handleMessages(pc)
 			go c.handleErrors(pc.Partition(), pc.Errors())
 		}
 	}()
+
+	// Expect to receive message and error handler "done" signals from each partition.
+	go waitForDoneSignals(c.messagesDoneChan, &c.doneWg, c.logger)
+	go waitForDoneSignals(c.errorsDoneChan, &c.doneWg, c.logger)
+}
+
+// waitForDoneSignals watches the doneChan for incoming "done" messages. If a message is received,
+// the doneWg WaitGroup is decremented via a call to Done().
+func waitForDoneSignals(doneChan <-chan string, doneWg *sync.WaitGroup, logger *zap.Logger) {
+	logger.Debug("Waiting for done signals")
+	for v := range doneChan {
+		logger.Debug("Received done signal", zap.String("msg", v))
+		doneWg.Done()
+	}
 }
 
 // Close closes the Consumer and underlying sarama consumer
 func (c *Consumer) Close() error {
-	c.partitionMapLock.Lock()
-	for _, p := range c.partitionIDToState {
-		c.closePartition(p.partitionConsumer)
-		p.wg.Wait()
-	}
-	c.partitionMapLock.Unlock()
-	c.deadlockDetector.close()
+	// Close the internal consumer, which will close each partition consumers' message and error channels.
 	c.logger.Info("Closing parent consumer")
-	return c.internalConsumer.Close()
+	err := c.internalConsumer.Close()
+
+	c.logger.Debug("Closing deadlock detector")
+	c.deadlockDetector.close()
+
+	c.logger.Debug("Waiting for messages and errors to be handled")
+	c.doneWg.Wait()
+
+	c.logger.Debug("Closing message and error done channels")
+	close(c.messagesDoneChan)
+	close(c.errorsDoneChan)
+
+	return err
 }
 
+// handleMessages handles incoming Kafka messages on a channel. Upon the closure of the message channel,
+// handleMessages will signal the messagesDoneChan to indicate the graceful shutdown of message handling is done.
 func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 	c.logger.Info("Starting message handler", zap.Int32("partition", pc.Partition()))
 	c.partitionMapLock.Lock()
 	c.partitionsHeld++
 	c.partitionsHeldGauge.Update(c.partitionsHeld)
-	wg := &c.partitionIDToState[pc.Partition()].wg
 	c.partitionMapLock.Unlock()
 	defer func() {
 		c.closePartition(pc)
-		wg.Done()
 		c.partitionMapLock.Lock()
 		c.partitionsHeld--
 		c.partitionsHeldGauge.Update(c.partitionsHeld)
 		c.partitionMapLock.Unlock()
+		c.messagesDoneChan <- "HandleMessages done"
 	}()
 
 	msgMetrics := c.newMsgMetrics(pc.Partition())
@@ -165,12 +185,13 @@ func (c *Consumer) closePartition(partitionConsumer sc.PartitionConsumer) {
 	c.logger.Info("Closed partition consumer", zap.Int32("partition", partitionConsumer.Partition()))
 }
 
+// handleErrors handles incoming Kafka consumer errors on a channel. Upon the closure of the error channel,
+// handleErrors will signal the errorsDoneChan to indicate the graceful shutdown of error handling is done.
 func (c *Consumer) handleErrors(partition int32, errChan <-chan *sarama.ConsumerError) {
 	c.logger.Info("Starting error handler", zap.Int32("partition", partition))
-	c.partitionMapLock.Lock()
-	wg := &c.partitionIDToState[partition].wg
-	c.partitionMapLock.Unlock()
-	defer wg.Done()
+	defer func() {
+		c.errorsDoneChan <- "HandleErrors done"
+	}()
 
 	errMetrics := c.newErrMetrics(partition)
 	for err := range errChan {

--- a/cmd/ingester/app/consumer/consumer.go
+++ b/cmd/ingester/app/consumer/consumer.go
@@ -51,9 +51,7 @@ type Consumer struct {
 	partitionsHeld      int64
 	partitionsHeldGauge metrics.Gauge
 
-	messagesDoneChan chan string
-	errorsDoneChan   chan string
-	doneWg           sync.WaitGroup
+	doneWg sync.WaitGroup
 }
 
 type consumerState struct {
@@ -71,8 +69,6 @@ func New(params Params) (*Consumer, error) {
 		deadlockDetector:    deadlockDetector,
 		partitionIDToState:  make(map[int32]*consumerState),
 		partitionsHeldGauge: partitionsHeldGauge(params.MetricsFactory),
-		messagesDoneChan:    make(chan string),
-		errorsDoneChan:      make(chan string),
 	}, nil
 }
 
@@ -92,20 +88,6 @@ func (c *Consumer) Start() {
 			go c.handleErrors(pc.Partition(), pc.Errors())
 		}
 	}()
-
-	// Expect to receive message and error handler "done" signals from each partition.
-	go waitForDoneSignals(c.messagesDoneChan, &c.doneWg, c.logger)
-	go waitForDoneSignals(c.errorsDoneChan, &c.doneWg, c.logger)
-}
-
-// waitForDoneSignals watches the doneChan for incoming "done" messages. If a message is received,
-// the doneWg WaitGroup is decremented via a call to Done().
-func waitForDoneSignals(doneChan <-chan string, doneWg *sync.WaitGroup, logger *zap.Logger) {
-	logger.Debug("Waiting for done signals")
-	for v := range doneChan {
-		logger.Debug("Received done signal", zap.String("msg", v))
-		doneWg.Done()
-	}
 }
 
 // Close closes the Consumer and underlying sarama consumer
@@ -120,15 +102,10 @@ func (c *Consumer) Close() error {
 	c.logger.Debug("Waiting for messages and errors to be handled")
 	c.doneWg.Wait()
 
-	c.logger.Debug("Closing message and error done channels")
-	close(c.messagesDoneChan)
-	close(c.errorsDoneChan)
-
 	return err
 }
 
-// handleMessages handles incoming Kafka messages on a channel. Upon the closure of the message channel,
-// handleMessages will signal the messagesDoneChan to indicate the graceful shutdown of message handling is done.
+// handleMessages handles incoming Kafka messages on a channel
 func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 	c.logger.Info("Starting message handler", zap.Int32("partition", pc.Partition()))
 	c.partitionMapLock.Lock()
@@ -141,7 +118,7 @@ func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 		c.partitionsHeld--
 		c.partitionsHeldGauge.Update(c.partitionsHeld)
 		c.partitionMapLock.Unlock()
-		c.messagesDoneChan <- "HandleMessages done"
+		c.doneWg.Done()
 	}()
 
 	msgMetrics := c.newMsgMetrics(pc.Partition())
@@ -185,13 +162,10 @@ func (c *Consumer) closePartition(partitionConsumer sc.PartitionConsumer) {
 	c.logger.Info("Closed partition consumer", zap.Int32("partition", partitionConsumer.Partition()))
 }
 
-// handleErrors handles incoming Kafka consumer errors on a channel. Upon the closure of the error channel,
-// handleErrors will signal the errorsDoneChan to indicate the graceful shutdown of error handling is done.
+// handleErrors handles incoming Kafka consumer errors on a channel
 func (c *Consumer) handleErrors(partition int32, errChan <-chan *sarama.ConsumerError) {
 	c.logger.Info("Starting error handler", zap.Int32("partition", partition))
-	defer func() {
-		c.errorsDoneChan <- "HandleErrors done"
-	}()
+	defer c.doneWg.Done()
 
 	errMetrics := c.newErrMetrics(partition)
 	for err := range errChan {

--- a/cmd/ingester/app/consumer/consumer_test.go
+++ b/cmd/ingester/app/consumer/consumer_test.go
@@ -68,7 +68,7 @@ func (s partitionConsumerWrapper) Topic() string {
 	return s.topic
 }
 
-func newSaramaClusterConsumer(saramaPartitionConsumer sarama.PartitionConsumer) *kmocks.Consumer {
+func newSaramaClusterConsumer(saramaPartitionConsumer sarama.PartitionConsumer, mc *smocks.PartitionConsumer) *kmocks.Consumer {
 	pcha := make(chan cluster.PartitionConsumer, 1)
 	pcha <- &partitionConsumerWrapper{
 		topic:             topic,
@@ -77,27 +77,26 @@ func newSaramaClusterConsumer(saramaPartitionConsumer sarama.PartitionConsumer) 
 	}
 	saramaClusterConsumer := &kmocks.Consumer{}
 	saramaClusterConsumer.On("Partitions").Return((<-chan cluster.PartitionConsumer)(pcha))
-	saramaClusterConsumer.On("Close").Return(nil)
+	saramaClusterConsumer.On("Close").Return(nil).Run(func(args mock.Arguments) {
+		mc.Close()
+	})
 	saramaClusterConsumer.On("MarkPartitionOffset", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(nil)
 	return saramaClusterConsumer
 }
 
 func newConsumer(
+	t *testing.T,
 	metricsFactory metrics.Factory,
 	topic string,
 	processor processor.SpanProcessor,
 	consumer consumer.Consumer) *Consumer {
 
 	logger, _ := zap.NewDevelopment()
-	return &Consumer{
-		metricsFactory:      metricsFactory,
-		logger:              logger,
-		internalConsumer:    consumer,
-		partitionIDToState:  make(map[int32]*consumerState),
-		partitionsHeldGauge: partitionsHeldGauge(metricsFactory),
-		deadlockDetector:    newDeadlockDetector(metricsFactory, logger, time.Second),
-
-		processorFactory: ProcessorFactory{
+	consumerParams := Params{
+		MetricsFactory:   metricsFactory,
+		Logger:           logger,
+		InternalConsumer: consumer,
+		ProcessorFactory: ProcessorFactory{
 			topic:          topic,
 			consumer:       consumer,
 			metricsFactory: metricsFactory,
@@ -106,6 +105,10 @@ func newConsumer(
 			parallelism:    1,
 		},
 	}
+
+	c, err := New(consumerParams)
+	require.NoError(t, err)
+	return c
 }
 
 func TestSaramaConsumerWrapper_MarkPartitionOffset(t *testing.T) {
@@ -136,7 +139,7 @@ func TestSaramaConsumerWrapper_start_Messages(t *testing.T) {
 	saramaPartitionConsumer, e := saramaConsumer.ConsumePartition(topic, partition, msgOffset)
 	require.NoError(t, e)
 
-	undertest := newConsumer(localFactory, topic, mp, newSaramaClusterConsumer(saramaPartitionConsumer))
+	undertest := newConsumer(t, localFactory, topic, mp, newSaramaClusterConsumer(saramaPartitionConsumer, mc))
 
 	undertest.partitionIDToState = map[int32]*consumerState{
 		partition: {
@@ -202,7 +205,7 @@ func TestSaramaConsumerWrapper_start_Errors(t *testing.T) {
 	saramaPartitionConsumer, e := saramaConsumer.ConsumePartition(topic, partition, msgOffset)
 	require.NoError(t, e)
 
-	undertest := newConsumer(localFactory, topic, &pmocks.SpanProcessor{}, newSaramaClusterConsumer(saramaPartitionConsumer))
+	undertest := newConsumer(t, localFactory, topic, &pmocks.SpanProcessor{}, newSaramaClusterConsumer(saramaPartitionConsumer, mc))
 
 	undertest.Start()
 	mc.YieldError(errors.New("Daisy, Daisy"))
@@ -238,7 +241,7 @@ func TestHandleClosePartition(t *testing.T) {
 	saramaPartitionConsumer, e := saramaConsumer.ConsumePartition(topic, partition, msgOffset)
 	require.NoError(t, e)
 
-	undertest := newConsumer(metricsFactory, topic, mp, newSaramaClusterConsumer(saramaPartitionConsumer))
+	undertest := newConsumer(t, metricsFactory, topic, mp, newSaramaClusterConsumer(saramaPartitionConsumer, mc))
 	undertest.deadlockDetector = newDeadlockDetector(metricsFactory, undertest.logger, 200*time.Millisecond)
 	undertest.Start()
 	defer undertest.Close()

